Add new public key token provider (tokens survive password change)

rullzer opened the pull request:
For #9441 

todo:
- [x] Finish unit tests

rullzer committed:
Abstract the Provider via a manager

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add new authtoken v2 columns to the authtoken table

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Initial PKT implementation

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add first tests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Just pass uid to the Token stuff

We don't have user objects in the code everywhere

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
SetPassword on PublicKeyTokens

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add publickey provider to manager

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Actually convert the token

* When getting the token
* When rotating the token

* Also store the encrypted password as base64 to avoid weird binary
stuff

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Generate the new publicKey tokens by default!

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add more tests

* Add a lot of tests
* Fixes related to those tests
* Fix tests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

ChristophWurst commented on lib/private/Authentication/Token/DefaultToken.php:
```diff
@@ -85,6 +89,9 @@ public function __construct() {
 		$this->addType('expires', 'int');
+		$this->addType('version', 'int');
+
+		$this->setVersion(1);
```
this will mark the `version` field as updated, although it won't actually be updated in many cases. This *could* cause unnecessary updates https://github.com/nextcloud/server/blob/a00cb2c5d73e598d5800176fafa6209675126b7e/lib/public/AppFramework/Db/Entity.php#L106

You might want to solve this differently.

ChristophWurst commented on lib/private/Authentication/Token/DefaultTokenMapper.php:
```diff
@@ -50,8 +49,8 @@ public function invalidate(string $token) {
-			->where($qb->expr()->eq('token', $qb->createParameter('token')))
-			->setParameter('token', $token)
+			->where($qb->expr()->eq('token', $qb->createNamedParameter($token, IQueryBuilder::PARAM_STR)))
+			->andWhere($qb->expr()->eq('version', $qb->createNamedParameter(1, IQueryBuilder::PARAM_INT)))
```
I'd appreciate using a constant for the version number :wink:

ChristophWurst commented on lib/private/Authentication/Token/IProvider.php:
```diff
@@ -92,10 +91,10 @@ public function invalidateToken(string $token);
 	 * @param int $id
 	 */
-	public function invalidateTokenById(IUser $user, int $id);
+	public function invalidateTokenById(string $uid, int $id);
```
I usually prefer to work with `IUser` objects on higher-level services over stringly-typed UIDs. What's the reasoning for choosing strings over rich objects?

rullzer replied:
Yes. Because we want to update all other tokens when we cdhange a password. I would need to inject the user manager to get the user object. Because the Tokens only have the uid.

ChristophWurst commented on lib/private/Authentication/Token/Manager.php:
```diff
@@ -0,0 +1,233 @@
+	 * @throws InvalidTokenException
+	 */
+	public function updateToken(IToken $token) {
+		if ($token instanceof DefaultToken) {
```
IMO it would make the code a lot easier if there was a private method `getProvider($token)`.

Like

```php
$this->getProvider($token)->updateToken($token);
```

where `getProvider` automatically throws the `InvalidTokenException` for unkown token types.

esp. since it's used many times in this class

ChristophWurst commented on lib/private/Authentication/Token/PublicKeyToken.php:
```diff
@@ -0,0 +1,216 @@
+		$this->addType('privateKey', 'string');
+		$this->addType('version', 'int');
+
+		$this->setVersion(2);
```
same as above, this could be problematic for the dirty checking of the entity/mapper logic

ChristophWurst commented on lib/private/Authentication/Token/PublicKeyTokenMapper.php:
```diff
@@ -0,0 +1,172 @@
+		$qb = $this->db->getQueryBuilder();
+		$qb->delete('authtoken')
+			->where($qb->expr()->eq('token', $qb->createNamedParameter($token)))
+			->andWhere($qb->expr()->eq('version', $qb->createNamedParameter(2, IQueryBuilder::PARAM_INT)))
```
again, I would appreciate a constant for the token version

ChristophWurst commented on lib/private/Authentication/Token/PublicKeyTokenProvider.php:
```diff
@@ -0,0 +1,320 @@
+		}
+
+		$password = null;
+		if (!is_null($token->getPassword())) {
```
oh, I though you were in team `!== null` :see_no_evil:

rullzer replied:
Copy paste ;)

ChristophWurst commented on lib/private/Authentication/Token/PublicKeyTokenProvider.php:
```diff
@@ -0,0 +1,320 @@
+			throw new InvalidTokenException();
+		}
+
+		// When changeing passwords all temp tokens are deleted
```
typo

ChristophWurst commented on lib/private/Files/ObjectStore/S3ConnectionTrait.php:
```diff
@@ -98,6 +98,7 @@ protected function getConnection() {
 			$options['signature_version'] = 'v2';
 		}
 		$this->connection = new S3Client($options);
+		$this->connection->getCredentials();
```
not sure what you're trying to sneak into our code base here

rullzer replied:
good question...

ChristophWurst reviewed (dismissed):
Some early feedback, as requested :wink:

rullzer committed:
Use constant for token version

And don't set the version in the constructor. That would possible cause
to many updates.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add getProvider helper function

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
This is ready for review and testing.

blizzz commented on lib/private/Authentication/Token/Manager.php:
```diff
@@ -0,0 +1,215 @@
+		$provider->updateTokenActivity($token);
+	}
+
+	public function getTokenByUser(string $uid): array {
```
the mixture of provided phpdoc and missing phpdoc… here it would be useful to doc IToken[]

blizzz commented on lib/private/Authentication/Token/Manager.php:
```diff
@@ -0,0 +1,215 @@
+	 * @throws InvalidTokenException
+	 * @return IToken
+	 */
+	public function getTokenById(int $tokenId): IToken {
```
besides by the type, the signature looks the same as in `getToken`. I guess it is database ID vs. token, um, value? Perhaps we can get the naming more clear in either this method or the upper one?

rullzer replied:
Yeah I don't like it either. But this was basically just copy pasted from the current implementation.

blizzz replied:
see, the IProvider is not in public namespace, so … but yeah, :see_no_evil: :checkered_flag:

ChristophWurst replied:
Yup, it's confusing.

blizzz commented on lib/private/Authentication/Token/PublicKeyToken.php:
```diff
@@ -0,0 +1,216 @@
+	}
+
+	public function setScope($scope) {
+		if (\is_array($scope)) {
```
\ shouldn't be necessary

blizzz reviewed (commented):
(tbc)

rullzer committed:
Comments

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
@blizzz addressed most. Lets get this in :)

blizzz commented on lib/private/Authentication/Token/Manager.php:
```diff
@@ -0,0 +1,219 @@
+		if ($token instanceof PublicKeyToken) {
+			return $this->publicKeyTokenProvider;
+		}
+		throw new InvalidTokenException();
```
annotate? would make IDE more calm ;)

blizzz commented on lib/private/Authentication/Token/PublicKeyTokenProvider.php:
```diff
@@ -0,0 +1,321 @@
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\IConfig;
+use OCP\ILogger;
+use OCP\IUser;
```
:fire:

blizzz commented on lib/private/Authentication/Token/PublicKeyTokenMapper.php:
```diff
@@ -0,0 +1,172 @@
+	 * where a high number of (session) tokens is generated
+	 *
+	 * @param string $uid
+	 * @return DefaultToken[]
```
PublicKeyToken[]

blizzz reviewed (approved):
just some :lipstick: remarks, looks good and works otherwise :+1:

rullzer commented on lib/private/Authentication/Token/PublicKeyTokenProvider.php:
```diff
@@ -0,0 +1,321 @@
+		$secret = $this->config->getSystemValue('secret');
+		try {
+			return $this->crypto->decrypt($cipherText, $token . $secret);
+		} catch (\Exception $ex) {
```
I just want to catch it all. Better to throw an invalid token exception in the login and have it fail then do unexpected stuff.

blizzz replied:
ok

ChristophWurst commented on lib/private/Authentication/Token/PublicKeyTokenProvider.php:
```diff
@@ -0,0 +1,320 @@
+	/** @var IConfig */
+	private $config;
+
+	/** @var ILogger $logger */
```
nit: `$logger` seems to be unnecessary, ie the doc comments are different to the ones above

blizzz replied:
not worth to wait another iteration, let's just get this out as by catch one next pr

ChristophWurst commented on lib/private/Authentication/Token/PublicKeyTokenProvider.php:
```diff
@@ -0,0 +1,320 @@
+		$pkToken->setExpires($defaultToken->getExpires());
+		$pkToken->setId($defaultToken->getId());
+
+		return $this->mapper->update($pkToken);
```
It seemed weird that you're using `update` for a new token, but then I realized it's the same table/row :+1:

ChristophWurst commented on version.php:
```diff
@@ -29,7 +29,7 @@
 // when updating major/minor version number.
 
-$OC_Version = array(14, 0, 0, 4);
+$OC_Version = array(14, 0, 0, 5);
```
note to myself: there will be conflicts with https://github.com/nextcloud/server/pull/9632/files#diff-5bbd3ccd25214f9956eea7e9f714bc08R32

ChristophWurst reviewed (approved):
Code looks good!
